### PR TITLE
Warn user that pre-release dependencies need to be explicitly listed in the Gemfile

### LINKED ISSUE #4657: Add an error message about pre-release dependencies
From https://github.com/bundler/bundler/issues/4340, https://github.com/bundler/bundler/issues/2938, https://github.com/bundler/bundler/issues/4337

It would help reduce user confusion if the error message generated in that case explicitly stated that pre-release dependencies need to be explicitly listed in the gemfile

### PR DESCRIPTION
**Pull request by b-ggs**
- [x] Add failing spec
- [x] Add warning

---

Closes #4657.
@segiddins @RochesterinNYC

### CODE REVIEW

**Comment by b-ggs** [comment]
So far, I've managed to recreate the issue in #4340. Will be figuring out when and under which conditions to display the warning.

**Comment by b-ggs** [comment]
Added the baseline implementation. So, the output would be:
```
boggs@hackbook.local ~/dev/test-bundler/4657/bundled_app
$ dbundle install
Fetching source index from file:/Users/boggs/dev/bundler/tmp/gems/remote2/
Resolving dependencies...
Bundler could not find compatible versions for gem "rspec-core":
  In Gemfile:
      guard-rspec (= 4.6) was resolved to 4.6.0, which depends on
        rspec (< 4.0, >= 2.99.0) was resolved to 2.99.0, which depends on
          rspec-core (~> 2.99.0)

      rspec-rails (= 3.5.0.beta1) was resolved to 3.5.0.beta1, which depends on
        rspec-core (= 3.5.0.beta1)

NOTE: Pre-release dependencies need to be explicitly listed in the Gemfile.
```

Thoughts on this? @segiddins @RochesterinNYC

File: spec/commands/install_spec.rb [hunk 1]
```diff
@@ -506,4 +506,31 @@
...
+        end
+      end
+
+      install_gemfile <<-G
+        source "file://#{gem_repo2}"
+        gem 'rspec-rails', '~> 3.5.0.beta1'
+        gem 'guard-rspec', '~> 4.6'
+      G
+
+      expect(out).to include("Pre-release dependencies need to be explicitly listed in the Gemfile.")
```
Comment (b-ggs) [inline spec/commands/install_spec.rb]: @segiddins Gotcha.

**Comment by b-ggs** [comment]
Ah, I seem to keep breaking unrelated specs here and there. 

```
Failures:
  1) bundle install with git sources gem install hooks runs pre-install hooks
     Failure/Error: expect(err).to eq_err("Ran pre-install hook: foo-1.0")
     
       expected "" to eq err "Ran pre-install hook: foo-1.0"
     
       Commands:
       $ /home/travis/.rvm/rubies/ruby-1.8.7-p374/bin/ruby -I/home/travis/build/bundler/bundler/lib:/home/travis/build/bundler/bundler/spec -r/home/travis/build/bundler/bundler/tmp/libs/install_hooks.rb -rsupport/hax /home/travis/build/bundler/bundler/exe/bundle install --no-color
     # ./spec/install/gemfile/git_spec.rb:875
Finished in 9 minutes 0 seconds (files took 2.67 seconds to load)
2036 examples, 1 failure, 9 pending
Failed examples:
rspec ./spec/install/gemfile/git_spec.rb:858 # bundle install with git sources gem install hooks runs pre-install hooks
```

Maybe a Travis fluke? Could you re-run the specs for this?

**Comment by b-ggs** [comment]
Aha! Looks like the specs finally passed haha.

**Comment by segiddins** [comment]
@homu delegate=RochesterinNYC

**Comment by RochesterinNYC** [comment]
@homu r+

**Comment by segiddins** [comment]
@homu retry

**Comment by indirect** [comment]
@homu retry

File: lib/bundler/resolver.rb [hunk 2]
```diff
@@ -61,6 +63,9 @@ def message
               o << "Could not find gem '#{conflict.requirement}' in any of the sources\n"
             end
           end
+          if has_prerelease_deps
+            o << "\nNOTE: Pre-release dependencies need to be explicitly listed in the Gemfile."
```
Comment (segiddins) [inline lib/bundler/resolver.rb]: maybe change this to `Please not that pre-release dependencies...` ?

**Comment by indirect** [comment]
@bundlerbot r+

**Comment by segiddins** [comment]
It turns out the given test for this change is an example of when the warning does _not_ need to be printed, as the conflict has nothing to do with the fact that `rspec-core` is a pre-release.

```
Bundler could not find compatible versions for gem "rspec-core":
  In Gemfile:
    guard-rspec (~> 4.6) was resolved to 4.6.0, which depends on
      rspec (< 4.0, >= 2.99.0) was resolved to 2.99.0, which depends on
        rspec-core (~> 2.99.0)

    rspec-rails (~> 3.5.0.beta1) was resolved to 3.5.0.beta1, which depends on
      rspec-core (= 3.5.0.beta1)

NOTE: Pre-release dependencies need to be explicitly listed in the Gemfile.
```

**Comment by deivid-rodriguez** [comment]
The original issue, https://github.com/bundler/bundler/issues/4657, got properly fixed in the resolver, so this warning shouldn't be needed. Closing!

**Pull request closed without merging**